### ISSUE TITLE
[clang-tidy] bugprone-fold-init-type check doesn't detect type mismatch when functor is provided

### ISSUE THREAD

**Issue by atom00** [llvm/llvm-project#85757]
When running bugprone-fold-init-type check on the following code clang-tidy doesn't warn about possible precision loss:
```
auto a = {65536LL, 65536LL,  65536LL};
return std::accumulate(std::begin(a), std::end(a), 1,  std::multiplies<>());
```

Removing std::multiplies makes the check work correctly.
